### PR TITLE
refactor: make class 'dns_resolver' as a singleton

### PR DESCRIPTION
**Pull request by GehaFearless**

### CODE REVIEW

File: src/runtime/rpc/dns_resolver.cpp [hunk 1]
```diff
@@ -141,4 +141,6 @@ rpc_address dns_resolver::resolve_address(const host_port &hp)
     }
 }
 
+dns_resolver::~dns_resolver() { _dns_cache.clear(); }
```
Comment (empiredan) [inline src/runtime/rpc/dns_resolver.cpp:144]: `_dns_cache` would be released automatically at the destruction.

File: src/runtime/test/host_port_test.cpp [hunk 2]
```diff
@@ -230,10 +230,9 @@ TEST(host_port_test, transfer_rpc_address)
 
 TEST(host_port_test, dns_resolver)
 {
-    dns_resolver resolver;
     {
         host_port hp("localhost", 8080);
-        auto addr = resolver.resolve_address(hp);
+        auto addr = dns_resolver::instance().resolve_address(hp);
```
Comment (empiredan) [inline src/runtime/test/host_port_test.cpp:235]:
```suggestion
        const auto &addr = dns_resolver::instance().resolve_address(hp);
```

File: src/runtime/test/host_port_test.cpp [hunk 3]
```diff
@@ -248,7 +247,7 @@ TEST(host_port_test, dns_resolver)
         host_port hp2("localhost", 8081);
         g_hp->set_leader(hp2);
 
-        auto addr_grp = resolver.resolve_address(hp_grp);
+        auto addr_grp = dns_resolver::instance().resolve_address(hp_grp);
```
Comment (empiredan) [inline src/runtime/test/host_port_test.cpp:250]:
```suggestion
        const auto &addr_grp = dns_resolver::instance().resolve_address(hp_grp);
```

File: src/runtime/test/host_port_test.cpp [hunk 4]
```diff
@@ -248,7 +247,7 @@ TEST(host_port_test, dns_resolver)
...
         auto g_addr = addr_grp.group_address();
```
Comment (empiredan) [inline src/runtime/test/host_port_test.cpp:251]: `rpc_group_address::count()` could also be declared as `const`:
```suggestion
        const auto * const g_addr = addr_grp.group_address();
```

**Review by empiredan — APPROVED** [review]

**Review by acelyc111 — APPROVED** [review]

**Pull request merged**